engineccl: add a method to verify batch repr integrity

danhhz opened the pull request:
VerifyBatchRepr asserts that all keys in a BatchRepr are between the
specified start and end keys and computes the enginepb.MVCCStats for it.

Notably, this is our first ccl'd c++ code. A new header (and namespace
internal) are introduced to allow the ccl tree to use code that we don't
want to be callable from go.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/13347)
<!-- Reviewable:end -->

danhhz committed:
engineccl: add a method to verify batch repr integrity

VerifyBatchRepr asserts that all keys in a BatchRepr are between the
specified start and end keys and computes the enginepb.MVCCStats for it.

Notably, this is our first ccl'd c++ code. A new header (and namespace
internal) are introduced to allow the ccl tree to use code that we don't
want to be callable from go.

danhhz commented:
@petermattis you already LGTM'd the impl of DBBatchReprVerify in https://github.com/cockroachdb/cockroach/pull/13104, so this is just to establish the mechanics of the c++ split

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 8 of 8 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/ccl/storageccl/engineccl/rocksdb.go, line 43 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13347#-KbvVMNZY0_79ikaXkzI:-KbvVMNZY0_79ikaXkzJ:b-prbt7o) ([raw file](https://github.com/cockroachdb/cockroach/blob/a203698ee49d5216f704972476785451c85a9443/pkg/ccl/storageccl/engineccl/rocksdb.go#L43)):*
> ```Go
> }
> 
> // TODO(dan): The following are all duplicated from storage/engine/rocksdb.go,
> ```

I've run into issues similar to this and don't know of any good solution.

A crazy idea: could we put these functions in their own file in storage/engine and then symlink that `.go` file into this package? 

---

*[pkg/ccl/storageccl/engineccl/rocksdb.go, line 88 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13347#-KbvUi8d6sZwSnrEFZB0:-KbvUi8d6sZwSnrEFZB1:b1is4hk) ([raw file](https://github.com/cockroachdb/cockroach/blob/a203698ee49d5216f704972476785451c85a9443/pkg/ccl/storageccl/engineccl/rocksdb.go#L88)):*
> ```Go
> }
> 
> func cStatsToGoStats(stats C.MVCCStatsResult, nowNanos int64) (enginepb.MVCCStats, error) {
> ```

I don't mind the duplication of the above functions too much, but it's unfortunate that we have another function that has to enumerate all of these fields. We should at least add a comment to `db.h` to point to the two functions that have to be updated whenever the MVCCStatsResult struct changes. 

---

*[pkg/storage/engine/rocksdb/db.cc, line 2378 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13347#-KbvXNKR2yJNcEoqjHvh:-KbvXNKR2yJNcEoqjHvi:b5xkyvc) ([raw file](https://github.com/cockroachdb/cockroach/blob/a203698ee49d5216f704972476785451c85a9443/pkg/storage/engine/rocksdb/db.cc#L2378)):*
> ```cpp
> 
> const rocksdb::Comparator* CockroachComparator() {
>   return (rocksdb::Comparator*)&kComparator;
> ```

Use `static_cast<rocksdb::Comparator*>(&kComparator)` (I think)

---

*[pkg/storage/engine/rocksdb/db_internal.h, line 23 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13347#-KbvWphWN7ikEAsn48Tk:-KbvWphWN7ikEAsn48Tl:bmtro0v) ([raw file](https://github.com/cockroachdb/cockroach/blob/a203698ee49d5216f704972476785451c85a9443/pkg/storage/engine/rocksdb/db_internal.h#L23)):*
> ```objc
> #include "rocksdb/write_batch_base.h"
> 
> namespace internal {
> ```

Comment that this is used to expose parts of this package to `rocksdbccl`. "Internal" isn't a great name for this: it's all internal, and the things in this namespace are  used more widely than the other parts, so maybe call it "shared"?

Actually, why is there even a namespace? Can't you just put these declarations in the top-level namespace?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13347#-:-KbvYEjss2IYKEXqGkTO:bnfp4nl)*
<!-- Sent from Reviewable.io -->

danhhz commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/ccl/storageccl/engineccl/rocksdb.go, line 43 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13347#-KbvVMNZY0_79ikaXkzI:-KbvbXXR6iWsshDcspZI:b338b2r) ([raw file](https://github.com/cockroachdb/cockroach/blob/a203698ee49d5216f704972476785451c85a9443/pkg/ccl/storageccl/engineccl/rocksdb.go#L43)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I've run into issues similar to this and don't know of any good solution.

A crazy idea: could we put these functions in their own file in storage/engine and then symlink that `.go` file into this package? 
</blockquote></details>

I like the symlink idea, but the package declaration needs to be different. Any ideas?

---

*[pkg/storage/engine/rocksdb/db_internal.h, line 23 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13347#-KbvWphWN7ikEAsn48Tk:-Kbvbz6jBrPU5R8w1vqS:b39iror) ([raw file](https://github.com/cockroachdb/cockroach/blob/a203698ee49d5216f704972476785451c85a9443/pkg/storage/engine/rocksdb/db_internal.h#L23)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Comment that this is used to expose parts of this package to `rocksdbccl`. "Internal" isn't a great name for this: it's all internal, and the things in this namespace are  used more widely than the other parts, so maybe call it "shared"?

Actually, why is there even a namespace? Can't you just put these declarations in the top-level namespace?
</blockquote></details>

The namespace is just me trying to organize the c++ code a bit. Am I overthinking? Would you prefer just putting this all in the top-level namespace?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13347)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/ccl/storageccl/engineccl/rocksdb.go, line 43 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13347#-KbvVMNZY0_79ikaXkzI:-Kbvcxf593pqc1wJ9kYy:bwriijf) ([raw file](https://github.com/cockroachdb/cockroach/blob/a203698ee49d5216f704972476785451c85a9443/pkg/ccl/storageccl/engineccl/rocksdb.go#L43)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

I like the symlink idea, but the package declaration needs to be different. Any ideas?
</blockquote></details>

The package name could be the same if we relax our convention that CCL packages end with "ccl". Alternately, we could use `go generate` instead of a symlink, to copy the file and change the package declaration (so the code would still be duplicated but CI wouldn't let them diverge). 

---

*[pkg/storage/engine/rocksdb/db_internal.h, line 23 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13347#-KbvWphWN7ikEAsn48Tk:-KbvdP1hYnKViSOCPq7J:b-6y4qb) ([raw file](https://github.com/cockroachdb/cockroach/blob/a203698ee49d5216f704972476785451c85a9443/pkg/storage/engine/rocksdb/db_internal.h#L23)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

The namespace is just me trying to organize the c++ code a bit. Am I overthinking? Would you prefer just putting this all in the top-level namespace?
</blockquote></details>

I prefer for (named) namespaces to correspond to file or directory structure; if you're going to put these functions in a separate namespace I'd move them into a separate `.cc` file (and separate files are probably enough organization even without the namespace). 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13347)*
<!-- Sent from Reviewable.io -->

danhhz commented:
Thanks for the review!

---

Review status: 4 of 10 files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/ccl/storageccl/engineccl/rocksdb.go, line 43 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13347#-KbvVMNZY0_79ikaXkzI:-KbvqV5VY_XYjAns_G7P:bysrbqa) ([raw file](https://github.com/cockroachdb/cockroach/blob/a203698ee49d5216f704972476785451c85a9443/pkg/ccl/storageccl/engineccl/rocksdb.go#L43)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

The package name could be the same if we relax our convention that CCL packages end with "ccl". Alternately, we could use `go generate` instead of a symlink, to copy the file and change the package declaration (so the code would still be duplicated but CI wouldn't let them diverge). 
</blockquote></details>

I'd like to keep the ccl convention for now. The go generate idea seems pretty workable, I added a mention of it to the TODO

---

*[pkg/ccl/storageccl/engineccl/rocksdb.go, line 88 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13347#-KbvUi8d6sZwSnrEFZB0:-Kbvqf3g8UxTY5stcQBi:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/a203698ee49d5216f704972476785451c85a9443/pkg/ccl/storageccl/engineccl/rocksdb.go#L88)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I don't mind the duplication of the above functions too much, but it's unfortunate that we have another function that has to enumerate all of these fields. We should at least add a comment to `db.h` to point to the two functions that have to be updated whenever the MVCCStatsResult struct changes. 
</blockquote></details>

Done

---

*[pkg/storage/engine/rocksdb/db.cc, line 2378 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13347#-KbvXNKR2yJNcEoqjHvh:-KbvqgdgcErK_S1OIrFb:bkj1adp) ([raw file](https://github.com/cockroachdb/cockroach/blob/a203698ee49d5216f704972476785451c85a9443/pkg/storage/engine/rocksdb/db.cc#L2378)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Use `static_cast<rocksdb::Comparator*>(&kComparator)` (I think)
</blockquote></details>

Turns out I didn't need a cast at all. I must have been doing something dumb at some intermediate point.

---

*[pkg/storage/engine/rocksdb/db_internal.h, line 23 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13347#-KbvWphWN7ikEAsn48Tk:-Kbvqu3xtTCEmR6fD0R4:beayt4h) ([raw file](https://github.com/cockroachdb/cockroach/blob/a203698ee49d5216f704972476785451c85a9443/pkg/storage/engine/rocksdb/db_internal.h#L23)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I prefer for (named) namespaces to correspond to file or directory structure; if you're going to put these functions in a separate namespace I'd move them into a separate `.cc` file (and separate files are probably enough organization even without the namespace). 
</blockquote></details>

Makes sense. Deleted the top-level namespace

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13347#-:-Kbvr012WtIKYGjDnUHv:b-9qdoz3)*
<!-- Sent from Reviewable.io -->
